File: #0238.product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {

        int size = nums.size();
        vector<int> result(size, 1);
        for (int i = 1; i < size; ++i) {
            result[i] = result[i - 1] * nums[i - 1];
        }

        int accumulate = 1;
        for (int i = size - 1; i >= 0; --i) {
            result[i] *= accumulate;
            accumulate *= nums[i];
        }
        return result;
    }
};

File: #0238.product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    const length = nums.length;
    const result = [];
    for (let i = 0; i < length; i++) {
        result[i] = (result[i - 1] ?? 1) * (nums[i - 1] ?? 1);
    }

    let rightProduct = 1;
    for (let i = length - 1; i >= 0; i--) {
        result[i] *= rightProduct;
        rightProduct *= nums[i];
    }

    return result;
};
